framework/src/meshgenerators/FancyExtruderGenerator.C

```diff
@@ -0,0 +1,547 @@
+      "subdomain_swaps",
+      "For each row, every two entries are interpreted as a pair of "
+      "'from' and 'to' to remap the subdomains for that elevation");
+
+  params.addRequiredParam<Point>(
+      "direction",
+      "A vector that points in the direction to extrude (note, this will be "
+      "normalized internally - so don't worry about it here)");
+
+  params.addParam<boundary_id_type>(
+      "top_boundary",
+      "The boundary ID to set on the top boundary.  If ommitted one will be generated.");
+
+  params.addParam<boundary_id_type>(
+      "bottom_boundary",
+      "The boundary ID to set on the bottom boundary.  If omitted one will be generated.");
+
+  return params;
+}
+
+FancyExtruderGenerator::FancyExtruderGenerator(const InputParameters & parameters)
+  : MeshGenerator(parameters),
+    _input(getMesh("input")),
+    _heights(getParam<std::vector<Real>>("heights")),
+    _num_layers(getParam<std::vector<unsigned int>>("num_layers")),
+    _subdomain_swaps(getParam<std::vector<std::vector<subdomain_id_type>>>("subdomain_swaps")),
+    _direction(getParam<Point>("direction")),
+    _has_top_boundary(isParamValid("top_boundary")),
+    _top_boundary(isParamValid("top_boundary") ? getParam<boundary_id_type>("top_boundary") : 0),
+    _has_bottom_boundary(isParamValid("bottom_boundary")),
+    _bottom_boundary(isParamValid("bottom_boundary") ? getParam<boundary_id_type>("bottom_boundary")
+                                                     : 0)
+{
+  // Normalize it
+  _direction /= _direction.norm();
+
+  const auto num_elevations = _heights.size();
+
+  if (_num_layers.size() != num_elevations)
+    mooseError("The length of 'heights' and 'num_layers' must be the same in ", name());
+
+  if (_subdomain_swaps.size() && (_subdomain_swaps.size() != num_elevations))
+    mooseError("If specified, 'subdomain_swaps' must be the same length as 'heights' in ", name());
+
+  _subdomain_swap_pairs.resize(_subdomain_swaps.size());
+
+  // Reprocess the subdomain swaps to make pairs out of them so they are easier to use
+  for (MooseIndex(_subdomain_swaps) i = 0; i < _subdomain_swaps.size(); i++)
+  {
+    const auto & elevation_swaps = _subdomain_swaps[i];
+    auto & elevation_swap_pairs = _subdomain_swap_pairs[i];
+
+    for (MooseIndex(elevation_swaps) i = 0; i < elevation_swaps.size(); i += 2)
+      elevation_swap_pairs[elevation_swaps[i]] = elevation_swaps[i + 1];
+  }
+}
+
+std::unique_ptr<MeshBase>
+FancyExtruderGenerator::generate()
+{
+  // Note: bulk of this code originally from libmesh mesh_modification.C
+  // Original copyright: Copyright (C) 2002-2019 Benjamin S. Kirk, John W. Peterson, Roy H. Stogner
+  // Original license is LGPL so it can be used here.
+
+  auto mesh = _mesh->buildMeshBaseObject();
+
+  std::unique_ptr<MeshBase> input = std::move(_input);
+
+  unsigned int total_num_layers = std::accumulate(_num_layers.begin(), _num_layers.end(), 0);
```

Comment:
This is just bizarre. Is this line of code really in libMesh?! Why not `size()`?

Reply:
Ah - nevermind - I see what this is doing now.

Reply:
This is my line of code - glad you saw what it does :-)